Validate attributes

anakrish opened the pull request:
Add validation for attributes supported by oeedger8r.
Lockdown with tests.

Signed-off-by: Anand Krishnamoorthi <anakrish@microsoft.com>

mingweishih commented on test/attributes/attributes.edl:
```diff
@@ -0,0 +1,138 @@
+// Licensed under the MIT License.
+
+enclave
+{
```
Can we have tests for struct as well?

anakrish replied:
Done.

mingweishih commented on parser.cpp:
```diff
@@ -634,53 +664,66 @@ Attrs* Parser::parse_attributes(bool fcn)
+        AttrTok atok = check_attribute(t);
+
+        // Only count and size attributes are valid for struct properties.
+        if (!fcn && (atok != TokCount && atok != TokSize))
```
Does this mean we still allow size and count for `enum`?

mingweishih replied:
Should we also support `string`?

anakrish replied:
> Does this mean we still allow size and count for `enum`?

For enums it is a syntax error to specify attributes. Not supported by the grammar.

For unions, we check and error out if attributes have been specified.
Added tests.

anakrish replied:
> Should we also support `string`?

That is a feature that we can add in the future. Maybe before that we could switch to a meta data based marshaling approach.

mingweishih commented:
For the case of adding attribute to the edl-annotated struct, it's required to have the `count` attribute be defined. Can we also error out the case that `count` is missing?

For example, we should error out the following case.

```
  struct DeepCopy
  {
    size_t size;
    [size=size] char *buf;
  };

...

void ocall_test([in, out] DeepCopy *dc)
```

 Otherwise, there will be compilation error: 
```
/home/ssh_office/workspace/oeedger8r-cpp/build/test/null_check/enc/null_check_t.c:1020:38: error: expected expression before ‘;’ token
         for (size_t _i_1 = 0; _i_1 < ; _i_1++)`.
```

anakrish commented:
> For the case of adding attribute to the edl-annotated struct, it's required to have the `count` attribute be defined. Can we also error out the case that `count` is missing?
> 
> For example, we should error out the following case.
> 
> ```
>   struct DeepCopy
>   {
>     size_t size;
>     [size=size] char *buf;
>   };
> 
> ...
> 
> void ocall_test([in, out] DeepCopy *dc)
> ```
> 
> Otherwise, there will be a compilation error:
> 
> ```
> /home/ssh_office/workspace/oeedger8r-cpp/build/test/null_check/enc/null_check_t.c:1020:38: error: expected expression before ‘;’ token
>          for (size_t _i_1 = 0; _i_1 < ; _i_1++)`.
> ```

@mingweishih That looks like a bug as opposed to a missing validation. Let me address that in a separate PR.

anakrish commented:
@mingweishih I have addressed all PR feedback. Can you take another look? 
Note: Added `in_function_` and `in_struct_.` parser state variables instead of passing flags.

mingweishih reviewed (approved):
LGTM. Thanks!
